File: app/src/main/java/com/nearsoft/nearbooks/models/BookModel.java
package com.nearsoft.nearbooks.models;

import android.content.Context;
import android.databinding.ViewDataBinding;
import android.text.TextUtils;

import com.nearsoft.nearbooks.NearbooksApplication;
import com.nearsoft.nearbooks.models.realm.Book;
import com.nearsoft.nearbooks.models.realm.Borrow;
import com.nearsoft.nearbooks.models.view.User;
import com.nearsoft.nearbooks.util.ErrorUtil;
import com.nearsoft.nearbooks.util.ViewUtil;
import com.nearsoft.nearbooks.ws.BookService;
import com.nearsoft.nearbooks.ws.bodies.GoogleBookBody;
import com.nearsoft.nearbooks.ws.bodies.RequestBody;
import com.nearsoft.nearbooks.ws.responses.AvailabilityResponse;
import com.nearsoft.nearbooks.ws.responses.MessageResponse;

import java.util.List;

import io.realm.Case;
import io.realm.Realm;
import io.realm.RealmResults;
import io.realm.Sort;
import retrofit2.Response;
import rx.Observable;
import rx.Subscriber;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Model to handle book actions.
 * Created by epool on 11/18/15.
 */
public class BookModel {

    private static BookService mBookService = NearbooksApplication.Companion
            .applicationComponent()
            .providesBookService();

    public static void cacheBooks(final List<Book> books) {
        if (books == null || books.isEmpty()) return;

        Realm realm = Realm.getDefaultInstance();
        realm.executeTransaction(r -> {
            r.delete(Book.class);
            r.copyToRealmOrUpdate(books);
        });
        realm.close();
    }

    public static Book findByBookId(String bookId) {
        Realm realm = Realm.getDefaultInstance();
        Book book = realm.where(Book.class)
                .equalTo("id", bookId)
                .findFirst();
        realm.close();
        return book;
    }

    public static RealmResults<Book> getAllBooks(Realm realm) {
        return realm.allObjects(Book.class).sort(Book.TITLE, Sort.ASCENDING);
    }

    public static RealmResults<Book> getBooksByQuery(Realm realm, String query) {
        if (TextUtils.isEmpty(query)) {
            return getAllBooks(realm);
        }

        return realm.where(Book.class)
                .contains(Book.ID, query, Case.INSENSITIVE)
                .or()
                .contains(Book.TITLE, query, Case.INSENSITIVE)
                .or()
                .contains(Book.AUTHOR, query, Case.INSENSITIVE)
                .or()
                .contains(Book.RELEASE_YEAR, query, Case.INSENSITIVE)
                .findAllSorted(Book.TITLE, Sort.ASCENDING);
    }

    public static Observable<Response<AvailabilityResponse>> checkBookAvailability(String bookId) {
        Observable<Response<AvailabilityResponse>> observable
                = mBookService.getBookAvailability(bookId + "-0");
        return observable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io());
    }

    public static Observable<Response<Borrow>> requestBookToBorrow(User user, String qrCode) {
        RequestBody requestBody = new RequestBody();
        requestBody.setQrCode(qrCode);
        requestBody.setUserEmail(user.getEmail());
        Observable<Response<Borrow>> observable = mBookService.requestBookToBorrow(requestBody);
        return observable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io());
    }

    public static Subscription doBookCheckIn(final ViewDataBinding binding, User user,
                                             String codeQr) {
        final Context context = binding.getRoot().getContext();
        RequestBody requestBody = new RequestBody();
        requestBody.setQrCode(codeQr);
        requestBody.setUserEmail(user.getEmail());
        Observable<Response<MessageResponse>> observable = mBookService.checkInBook(requestBody);
        return observable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(new Subscriber<Response<MessageResponse>>() {
                    @Override
                    public void onCompleted() {
                    }

                    @Override
                    public void onError(Throwable t) {
                        ViewUtil.showSnackbarMessage(binding, t.getLocalizedMessage());
                    }

                    @Override
                    public void onNext(Response<MessageResponse> response) {
                        if (response.isSuccessful()) {
                            MessageResponse messageResponse = response.body();
                            ViewUtil.showSnackbarMessage(binding, messageResponse.getMessage());
                        } else {
                            MessageResponse messageResponse = ErrorUtil
                                    .parseError(MessageResponse.class, response);
                            if (messageResponse != null) {
                                ViewUtil.showSnackbarMessage(binding, messageResponse.getMessage());
                            } else {
                                ViewUtil.showSnackbarMessage(binding,
                                        ErrorUtil.getGeneralExceptionMessage(context,
                                                response.code()));
                            }
                        }
                    }
                });
    }

    public static Subscription doBookCheckOut(final ViewDataBinding binding, User user,
                                              String codeQr) {
        final Context context = binding.getRoot().getContext();
        RequestBody requestBody = new RequestBody();
        requestBody.setQrCode(codeQr);
        requestBody.setUserEmail(user.getEmail());
        Observable<Response<MessageResponse>> observable = mBookService.checkOutBook(requestBody);
        return observable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(new Subscriber<Response<MessageResponse>>() {
                    @Override
                    public void onCompleted() {
                    }

                    @Override
                    public void onError(Throwable t) {
                        ViewUtil.showSnackbarMessage(binding, t.getLocalizedMessage());
                    }

                    @Override
                    public void onNext(Response<MessageResponse> response) {
                        if (response.isSuccessful()) {
                            MessageResponse messageResponse = response.body();
                            ViewUtil.showSnackbarMessage(binding, messageResponse.getMessage());
                        } else {
                            MessageResponse messageResponse = ErrorUtil
                                    .parseError(MessageResponse.class, response);
                            if (messageResponse != null) {
                                ViewUtil.showSnackbarMessage(binding, messageResponse.getMessage());
                            } else {
                                ViewUtil.showSnackbarMessage(binding,
                                        ErrorUtil.getGeneralExceptionMessage(context,
                                                response.code()));
                            }
                        }
                    }
                });
    }

    public static Observable<Response<MessageResponse>> registerNewBook(
            GoogleBookBody googleBookBody) {
        return mBookService.registerNewBook(googleBookBody)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io());
    }

}

File: app/src/main/java/com/nearsoft/nearbooks/models/BookModel.kt
package com.nearsoft.nearbooks.models

import android.databinding.ViewDataBinding
import android.text.TextUtils
import com.nearsoft.nearbooks.NearbooksApplication
import com.nearsoft.nearbooks.models.realm.Book
import com.nearsoft.nearbooks.models.realm.Borrow
import com.nearsoft.nearbooks.models.view.User
import com.nearsoft.nearbooks.util.ErrorUtil
import com.nearsoft.nearbooks.util.ViewUtil
import com.nearsoft.nearbooks.ws.bodies.GoogleBookBody
import com.nearsoft.nearbooks.ws.bodies.RequestBody
import com.nearsoft.nearbooks.ws.responses.AvailabilityResponse
import com.nearsoft.nearbooks.ws.responses.MessageResponse
import io.realm.Case
import io.realm.Realm
import io.realm.RealmResults
import io.realm.Sort
import retrofit2.Response
import rx.Observable
import rx.Subscriber
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

/**
 * Model to handle book actions.
 * Created by epool on 11/18/15.
 */
object BookModel {

    private val sBookService = NearbooksApplication.applicationComponent().provideBookService()

    fun cacheBooks(books: List<Book>?) {
        if (books?.isEmpty() ?: true) return

        val realm = Realm.getDefaultInstance()
        realm.executeTransaction { r ->
            r.delete(Book::class.java)
            r.copyToRealmOrUpdate(books)
        }
        realm.close()
    }

    fun findByBookId(bookId: String): Book {
        val realm = Realm.getDefaultInstance()
        val book = realm.where(Book::class.java).equalTo("id", bookId).findFirst()
        realm.close()
        return book
    }

    fun getAllBooks(realm: Realm): RealmResults<Book> {
        return realm.where(Book::class.java).findAll().sort(Book.TITLE, Sort.ASCENDING)
    }

    fun getBooksByQuery(realm: Realm, query: String): RealmResults<Book> {
        if (TextUtils.isEmpty(query)) {
            return getAllBooks(realm)
        }

        return realm.where(Book::class.java).contains(Book.ID, query, Case.INSENSITIVE).or().contains(Book.TITLE, query, Case.INSENSITIVE).or().contains(Book.AUTHOR, query, Case.INSENSITIVE).or().contains(Book.RELEASE_YEAR, query, Case.INSENSITIVE).findAllSorted(Book.TITLE, Sort.ASCENDING)
    }

    fun checkBookAvailability(bookId: String): Observable<Response<AvailabilityResponse>> {
        val observable = sBookService.getBookAvailability(bookId + "-0")
        return observable.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).unsubscribeOn(Schedulers.io())
    }

    fun requestBookToBorrow(user: User, qrCode: String): Observable<Response<Borrow>> {
        val requestBody = RequestBody()
        requestBody.qrCode = qrCode
        requestBody.userEmail = user.email
        val observable = sBookService.requestBookToBorrow(requestBody)
        return observable.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).unsubscribeOn(Schedulers.io())
    }

    fun doBookCheckIn(binding: ViewDataBinding, user: User,
                      codeQr: String): Subscription {
        val context = binding.root.context
        val requestBody = RequestBody()
        requestBody.qrCode = codeQr
        requestBody.userEmail = user.email
        val observable = sBookService.checkInBook(requestBody)
        return observable.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).unsubscribeOn(Schedulers.io()).subscribe(object : Subscriber<Response<MessageResponse>>() {
            override fun onCompleted() {
            }

            override fun onError(t: Throwable) {
                ViewUtil.showSnackbarMessage(binding, t.message!!)
            }

            override fun onNext(response: Response<MessageResponse>) {
                if (response.isSuccessful) {
                    val messageResponse = response.body()
                    ViewUtil.showSnackbarMessage(binding, messageResponse.message!!)
                } else {
                    val messageResponse = ErrorUtil.parseError(MessageResponse::class.java, response)
                    if (messageResponse != null) {
                        ViewUtil.showSnackbarMessage(binding, messageResponse.message!!)
                    } else {
                        ViewUtil.showSnackbarMessage(binding,
                                ErrorUtil.getGeneralExceptionMessage(context, response.code())!!)
                    }
                }
            }
        })
    }

    fun doBookCheckOut(binding: ViewDataBinding, user: User,
                       codeQr: String): Subscription {
        val context = binding.root.context
        val requestBody = RequestBody()
        requestBody.qrCode = codeQr
        requestBody.userEmail = user.email
        val observable = sBookService.checkOutBook(requestBody)
        return observable.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).unsubscribeOn(Schedulers.io()).subscribe(object : Subscriber<Response<MessageResponse>>() {
            override fun onCompleted() {
            }

            override fun onError(t: Throwable) {
                ViewUtil.showSnackbarMessage(binding, t.message!!)
            }

            override fun onNext(response: Response<MessageResponse>) {
                if (response.isSuccessful) {
                    val messageResponse = response.body()
                    ViewUtil.showSnackbarMessage(binding, messageResponse.message!!)
                } else {
                    val messageResponse = ErrorUtil.parseError(MessageResponse::class.java, response)
                    if (messageResponse != null) {
                        ViewUtil.showSnackbarMessage(binding, messageResponse.message!!)
                    } else {
                        ViewUtil.showSnackbarMessage(binding,
                                ErrorUtil.getGeneralExceptionMessage(context, response.code())!!)
                    }
                }
            }
        })
    }

    fun registerNewBook(
            googleBookBody: GoogleBookBody): Observable<Response<MessageResponse>> {
        return sBookService.registerNewBook(googleBookBody).observeOn(AndroidSchedulers.mainThread()).subscribeOn(Schedulers.io())
    }

}
